### ISSUE TITLE
LANG CHAIN

### ISSUE THREAD

**Issue by pseudoimran1** [pseudoimran1/MAJOR-V-SEM-5#1]
IMRAN TAMBOLI 
ROLL NO : 19051

# translate_app.py (modified version)
import re
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
import torch

# Model names
ENGLISH_TO_HINDI = "Helsinki-NLP/opus-mt-en-hi"
HINDI_TO_ENGLISH = "Helsinki-NLP/opus-mt-hi-en"

# Split text into sentences
def split_sentences(text):
    sentences = re.split(r'(?<=[.!?])\s+', text.strip())
    return [s for s in sentences if s]

# Create manageable chunks
def chunk_sentences(sentences, max_chars=350):
    chunks, current_chunk = [], ""
    for sentence in sentences:
        if len(current_chunk) + len(sentence) + 1 <= max_chars:
            current_chunk = (current_chunk + " " + sentence).strip() if current_chunk else sentence
        else:
            chunks.append(current_chunk)
            current_chunk = sentence
    if current_chunk:
        chunks.append(current_chunk)
    return chunks

# Translate a chunked text
def translate_text(text, model, tokenizer, device):
    sentences = split_sentences(text)
    chunks = chunk_sentences(sentences)
    results = []
    for chunk in chunks:
        inputs = tokenizer([chunk], return_tensors="pt", padding=True).to(device)
        with torch.no_grad():
            output_tokens = model.generate(**inputs, max_length=500, num_beams=5, early_stopping=True)
        results.append(tokenizer.batch_decode(output_tokens, skip_special_tokens=True)[0])
    return " ".join(results)

# Load model based on translation direction
def load_translation_model(direction, device):
    if direction == "en-hi":
        tokenizer = AutoTokenizer.from_pretrained(ENGLISH_TO_HINDI)
        model = AutoModelForSeq2SeqLM.from_pretrained(ENGLISH_TO_HINDI).to(device)
    else:
        tokenizer = AutoTokenizer.from_pretrained(HINDI_TO_ENGLISH)
        model = AutoModelForSeq2SeqLM.from_pretrained(HINDI_TO_ENGLISH).to(device)
    return model, tokenizer

# Main program
def main():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print("="*50)
    print("🔥 Welcome to the English ↔️ Hindi Translator 🔥")
    print(f"Running on device: {device}")
    print("="*50)
    
    print("Select translation mode:")
    print("1) English → Hindi")
    print("2) Hindi → English")
    choice = input("Enter 1 or 2 (default 1): ").strip() or "1"
    direction = "en-hi" if choice == "1" else "hi-en"

    print("Loading translation model... (first run will download weights)")
    model, tokenizer = load_translation_model(direction, device)
    
    print("\nTranslator is ready! Type 'exit' to quit.\n")
    while True:
        text_input = input("Enter text to translate: ").strip()
        if text_input.lower() in ("exit", "quit"):
            print("Goodbye! 👋")
            break
        if not text_input:
            print("Please enter some text.")
            continue
        try:
            translation = translate_text(text_input, model, tokenizer, device)
            print("➡️ Translation:", translation, "\n")
        except RuntimeError as e:
            print("Runtime error (maybe GPU out-of-memory). Try shorter text or CPU.")
            print(str(e))
        except Exception as e:
            print("Translation failed:", str(e))

if __name__ == "__main__":
    main()
